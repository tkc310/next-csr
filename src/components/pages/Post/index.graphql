# @see
# - https://techblog.yahoo.co.jp/entry/2020121530052952/
# - https://vivit.hatenablog.com/entry/2020/05/22/193021

query PostIndex($id: Int!) {
  ...Post
}

fragment Post on Query {
  post(id: $id) {
    id
    title
    content
  }
}
