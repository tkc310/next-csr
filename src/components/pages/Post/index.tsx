import { FC } from 'react';
import { filter } from 'graphql-anywhere';
import {
  PostFragment,
  PostFragmentDoc,
  usePostIndexQuery,
} from '@/graphql/generated/pages/post';
import { useRouter } from 'next/dist/client/router';

type Props = {
  data: PostFragment;
};

// presentational component
export const PostPresentational: FC<Props> = ({ data }) => {
  const { post } = data;

  return post ? (
    <>
      <h1>Post Detail</h1>
      <h2>{post.title}</h2>
      <p>{post.content}</p>
    </>
  ) : (
    <p>Postが存在しません</p>
  );
};

// container component
const PostsContainer: FC = () => {
  const router = useRouter();
  const { post_id } = router.query;
  const id = parseInt(String(post_id));

  const { data, loading, error } = usePostIndexQuery({
    variables: { id: id },
  });

  if (loading) return <div>Loading</div>;
  if (error) return <div>Error</div>;

  return (
    <PostPresentational data={filter<PostFragment>(PostFragmentDoc, data)} />
  );
};

export default PostsContainer;
