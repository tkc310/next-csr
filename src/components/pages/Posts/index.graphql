# @see
# - https://techblog.yahoo.co.jp/entry/2020121530052952/
# - https://vivit.hatenablog.com/entry/2020/05/22/193021

query PostsIndex {
  ...Posts
}

fragment Posts on Query {
  posts {
    id
    title
    content
  }
}
