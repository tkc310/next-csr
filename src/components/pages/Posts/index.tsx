import { FC } from 'react';
import Link from 'next/link';
import { filter } from 'graphql-anywhere';
import {
  Post,
  PostsFragment,
  PostsFragmentDoc,
  usePostsIndexQuery,
} from '@/graphql/generated/pages/posts';

type Props = {
  data: PostsFragment;
};

// presentational component
export const PostsPresentational: FC<Props> = ({ data }) => {
  const { posts } = data;

  const castedPosts = posts as Post[];
  const postsNode = castedPosts.map((post) => (
    <li key={post.id} style={{ marginBottom: '16px' }}>
      <p style={{ fontWeight: 'bold' }}>{post.title}</p>
      <p>{post.content}</p>
      <div>
        <Link href={`/posts/${post?.id}`}>詳細</Link>
      </div>
    </li>
  ));

  return posts && posts.length ? (
    <>
      <ul>{postsNode}</ul>
    </>
  ) : (
    <p>Postが存在しません</p>
  );
};

// container component
const PostsContainer: FC = () => {
  const { data, loading, error } = usePostsIndexQuery();

  // TODO: ローディングコンポーネント
  if (loading) return <>Loading...</>;
  // TODO: graphqlのエラーハンドリング、500系のエラーページ
  if (error) return <>Error...</>;

  return (
    <PostsPresentational data={filter<PostsFragment>(PostsFragmentDoc, data)} />
  );
};

export default PostsContainer;
