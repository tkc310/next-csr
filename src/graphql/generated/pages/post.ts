import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
const defaultOptions = {};
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type CreateItem = {
  __typename?: 'CreateItem';
  item?: Maybe<Item>;
};

export type CreatePost = {
  __typename?: 'CreatePost';
  post?: Maybe<Post>;
};

export type Item = {
  __typename?: 'Item';
  id: Scalars['ID'];
  name: Scalars['String'];
  count: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  createPost?: Maybe<CreatePost>;
  createItem?: Maybe<CreateItem>;
};

export type MutationCreatePostArgs = {
  content?: Maybe<Scalars['String']>;
  title?: Maybe<Scalars['String']>;
};

export type MutationCreateItemArgs = {
  count?: Maybe<Scalars['Int']>;
  name?: Maybe<Scalars['String']>;
};

export type Post = {
  __typename?: 'Post';
  id: Scalars['ID'];
  title: Scalars['String'];
  content: Scalars['String'];
};

export type Query = {
  __typename?: 'Query';
  posts?: Maybe<Array<Maybe<Post>>>;
  post?: Maybe<Post>;
  items?: Maybe<Array<Maybe<Item>>>;
};

export type QueryPostArgs = {
  id?: Maybe<Scalars['Int']>;
};

export type PostIndexQueryVariables = Exact<{
  id: Scalars['Int'];
  editable: Scalars['Boolean'];
}>;

export type PostIndexQuery = { __typename?: 'Query' } & PostFragment;

export type PostFragment = { __typename?: 'Query' } & {
  post?: Maybe<
    { __typename?: 'Post' } & Pick<Post, 'id' | 'title' | 'content'>
  >;
};

export const PostFragmentDoc = gql`
  fragment Post on Query {
    post(id: $id) @include(if: $editable) {
      id
      title
      content
    }
  }
`;
export const PostIndexDocument = gql`
  query PostIndex($id: Int!, $editable: Boolean!) {
    ...Post
  }
  ${PostFragmentDoc}
`;

/**
 * __usePostIndexQuery__
 *
 * To run a query within a React component, call `usePostIndexQuery` and pass it any options that fit your needs.
 * When your component renders, `usePostIndexQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = usePostIndexQuery({
 *   variables: {
 *      id: // value for 'id'
 *      editable: // value for 'editable'
 *   },
 * });
 */
export function usePostIndexQuery(
  baseOptions: Apollo.QueryHookOptions<PostIndexQuery, PostIndexQueryVariables>
) {
  const options = { ...defaultOptions, ...baseOptions };
  return Apollo.useQuery<PostIndexQuery, PostIndexQueryVariables>(
    PostIndexDocument,
    options
  );
}
export function usePostIndexLazyQuery(
  baseOptions?: Apollo.LazyQueryHookOptions<
    PostIndexQuery,
    PostIndexQueryVariables
  >
) {
  const options = { ...defaultOptions, ...baseOptions };
  return Apollo.useLazyQuery<PostIndexQuery, PostIndexQueryVariables>(
    PostIndexDocument,
    options
  );
}
export type PostIndexQueryHookResult = ReturnType<typeof usePostIndexQuery>;
export type PostIndexLazyQueryHookResult = ReturnType<
  typeof usePostIndexLazyQuery
>;
export type PostIndexQueryResult = Apollo.QueryResult<
  PostIndexQuery,
  PostIndexQueryVariables
>;
