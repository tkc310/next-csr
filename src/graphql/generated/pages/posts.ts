import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
const defaultOptions = {};
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
};

export type CreateItem = {
  __typename?: 'CreateItem';
  item?: Maybe<Item>;
};

export type CreatePost = {
  __typename?: 'CreatePost';
  post?: Maybe<Post>;
};

export type Item = {
  __typename?: 'Item';
  id: Scalars['ID'];
  name: Scalars['String'];
  count: Scalars['Int'];
};

export type Mutation = {
  __typename?: 'Mutation';
  createPost?: Maybe<CreatePost>;
  createItem?: Maybe<CreateItem>;
};

export type MutationCreatePostArgs = {
  content?: Maybe<Scalars['String']>;
  title?: Maybe<Scalars['String']>;
};

export type MutationCreateItemArgs = {
  count?: Maybe<Scalars['Int']>;
  name?: Maybe<Scalars['String']>;
};

export type Post = {
  __typename?: 'Post';
  id: Scalars['ID'];
  title: Scalars['String'];
  content: Scalars['String'];
};

export type Query = {
  __typename?: 'Query';
  posts?: Maybe<Array<Maybe<Post>>>;
  post?: Maybe<Post>;
  items?: Maybe<Array<Maybe<Item>>>;
};

export type QueryPostArgs = {
  id?: Maybe<Scalars['Int']>;
};

export type PostsIndexQueryVariables = Exact<{ [key: string]: never }>;

export type PostsIndexQuery = { __typename?: 'Query' } & PostsFragment;

export type PostsFragment = { __typename?: 'Query' } & {
  posts?: Maybe<
    Array<
      Maybe<{ __typename?: 'Post' } & Pick<Post, 'id' | 'title' | 'content'>>
    >
  >;
};

export const PostsFragmentDoc = gql`
  fragment Posts on Query {
    posts {
      id
      title
      content
    }
  }
`;
export const PostsIndexDocument = gql`
  query PostsIndex {
    ...Posts
  }
  ${PostsFragmentDoc}
`;

/**
 * __usePostsIndexQuery__
 *
 * To run a query within a React component, call `usePostsIndexQuery` and pass it any options that fit your needs.
 * When your component renders, `usePostsIndexQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = usePostsIndexQuery({
 *   variables: {
 *   },
 * });
 */
export function usePostsIndexQuery(
  baseOptions?: Apollo.QueryHookOptions<
    PostsIndexQuery,
    PostsIndexQueryVariables
  >
) {
  const options = { ...defaultOptions, ...baseOptions };
  return Apollo.useQuery<PostsIndexQuery, PostsIndexQueryVariables>(
    PostsIndexDocument,
    options
  );
}
export function usePostsIndexLazyQuery(
  baseOptions?: Apollo.LazyQueryHookOptions<
    PostsIndexQuery,
    PostsIndexQueryVariables
  >
) {
  const options = { ...defaultOptions, ...baseOptions };
  return Apollo.useLazyQuery<PostsIndexQuery, PostsIndexQueryVariables>(
    PostsIndexDocument,
    options
  );
}
export type PostsIndexQueryHookResult = ReturnType<typeof usePostsIndexQuery>;
export type PostsIndexLazyQueryHookResult = ReturnType<
  typeof usePostsIndexLazyQuery
>;
export type PostsIndexQueryResult = Apollo.QueryResult<
  PostsIndexQuery,
  PostsIndexQueryVariables
>;
