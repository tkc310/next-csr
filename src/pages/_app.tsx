import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  createHttpLink,
} from '@apollo/client';
import { URL } from '@/constants';
import type { AppProps } from 'next/app';
import '@styles/globals.css';
import { FC } from 'react';

const link = createHttpLink({
  uri: `${URL.API}/graphql/`,
  credentials: 'same-origin',
});
const cache = new InMemoryCache();
const client = new ApolloClient({
  link,
  cache,
});

const App: FC<AppProps> = ({ Component, pageProps }) => {
  return (
    <ApolloProvider client={client}>
      <Component {...pageProps} />
    </ApolloProvider>
  );
};

export default App;
