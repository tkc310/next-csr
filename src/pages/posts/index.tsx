import { memo, Fragment, useEffect } from 'react';
import useStateIfMounted from '@/hooks/useSafeState';
import { usePostsQuery } from '@/graphql';
import type { TPost } from '@/types/props';

type TProps = {};

const defaultProps = {};

const Posts = ({}: TProps) => {
  const { data, loading, error } = usePostsQuery();

  if (loading) return <div>Loading</div>;
  if (error) return <div>Error</div>;

  return (
    <div>
      <h1>Posts</h1>
      {data?.posts
        ? data.posts.map((post) => (
            <Fragment key={post.id}>
              <h2>{post.title}</h2>
              <p>{post.content}</p>
            </Fragment>
          ))
        : null}
    </div>
  );
};

Posts.defaultProps = defaultProps;

export default memo(Posts);
