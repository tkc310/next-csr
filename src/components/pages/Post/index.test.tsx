import { getPage } from 'next-page-tester';
import { screen } from '@testing-library/react';
import { PostFragment } from '@/graphql/generated/pages/post';

describe('Post', () => {
  const data: PostFragment = {
    post: {
      __typename: 'Post',
      id: '1',
      title: 'title foo',
      content: 'content',
    },
  };

  it('renders without crashing', async () => {
    const { render } = await getPage({
      route: '/',
    });

    render();
    // expect(screen.getByText('title foo')).toBeInTheDocument();
  });
});
