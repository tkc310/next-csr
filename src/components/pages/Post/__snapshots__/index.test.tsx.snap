// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post renders without crashing 1`] = `
"<div id=\\"__next\\"><div>Loading</div></div>
"
`;

exports[`Post renders without crashing 2`] = `
"<div id=\\"__next\\">
  <h1>Post Detail</h1>
  <h2>titile foo</h2>
  <p>content foo</p>
</div>
"
`;
